File: other/prime_checker.hs
#!/usr/bin/env runhaskell
import System.Environment (getArgs)
import System.IO (hFlush, stdout)
import Text.Read (readMaybe)

-- 素数判定を行う関数
-- n: 判定したい正の整数
-- 戻り値: 素数の場合True、そうでなければFalse
isPrime :: Int -> Bool
isPrime n
    | n < 2 = False
    | n == 2 = True
    | even n = False
    | otherwise = not $ any (\x -> n `mod` x == 0) [3, 5 .. floor $ sqrt $ fromIntegral n]

-- 数値を処理する関数
-- number: 処理する数値
processNumber :: Int -> IO ()
processNumber number
    | number <= 0 = putStrLn "エラー: 正の整数を入力してください"
    | number > 10000000 = putStrLn "エラー: 10,000,000以下の数値を入力してください"
    | isPrime number = putStrLn $ show number ++ " は素数です"
    | otherwise = putStrLn $ show number ++ " は素数ではありません"

-- 文字列を整数に変換する安全な関数
safeRead :: String -> Maybe Int
safeRead = readMaybe

-- ユーザー入力を取得する関数
getUserInput :: IO (Maybe Int)
getUserInput = do
    putStr "判定したい数値を入力してください: "
    hFlush stdout
    input <- getLine
    return $ safeRead input

-- メイン処理
main :: IO ()
main = do
    args <- getArgs
    maybeNumber <- case args of
        [] -> getUserInput  -- ユーザー入力から取得
        (arg:_) -> return $ safeRead arg  -- コマンドライン引数から取得
    
    case maybeNumber of
        Just number -> processNumber number
        Nothing -> putStrLn "エラー: 有効な整数を入力してください"

File: other/prime_checker.m
function prime_checker(varargin)
    % 素数判定プログラム（MATLAB版）
    % 使用方法: prime_checker() または prime_checker(number)
    
    if nargin > 0
        % 引数として数値が渡された場合
        number = varargin{1};
        if ~isnumeric(number) || ~isscalar(number) || number ~= floor(number)
            fprintf('エラー: 有効な整数を入力してください\n');
            return;
        end
    else
        % ユーザー入力から取得
        number = input('判定したい数値を入力してください: ');
        if ~isnumeric(number) || ~isscalar(number) || number ~= floor(number)
            fprintf('エラー: 有効な整数を入力してください\n');
            return;
        end
    end
    
    processNumber(number);
end

function result = isPrime(n)
    % 素数判定を行う関数
    % 入力: n - 判定したい正の整数
    % 出力: 素数の場合true、そうでなければfalse
    
    if n < 2
        result = false;
        return;
    end
    
    if n == 2
        result = true;
        return;
    end
    
    if mod(n, 2) == 0
        result = false;
        return;
    end
    
    % 3から√nまでの奇数で割り切れるかチェック
    sqrt_n = floor(sqrt(n));
    for i = 3:2:sqrt_n
        if mod(n, i) == 0
            result = false;
            return;
        end
    end
    
    result = true;
end

function processNumber(number)
    % 数値を処理する関数
    % 入力: number - 処理する数値
    
    if number <= 0
        fprintf('エラー: 正の整数を入力してください\n');
        return;
    end
    
    if number > 10000000
        fprintf('エラー: 10,000,000以下の数値を入力してください\n');
        return;
    end
    
    if isPrime(number)
        fprintf('%d は素数です\n', number);
    else
        fprintf('%d は素数ではありません\n', number);
    end
end